# Data Science Academy

## Capítulo 10 - Manipulação de Dados com Pandas

### Preenchendo Valores ausnetes em DataFrames do Pandas

A função `fillna()`é usada para preencher os valores ausentes. A função oferece muitas opções. Podemos usar um valor específico, uma função agregada (por exemplo, a média) ou um valor anterior ou seguinte. No exemplo abaixo, usaremos a moda para preencher o NaN.

In [2]:
import pandas as pd
import numpy as np

In [5]:
# Importanto o dataset
df = pd.read_csv('dataset.csv')

In [6]:
df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         NaN  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [8]:
# Somando a quantidade de valores ausentes de cada coluna
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [9]:
# Extraímos a moda da coluna Quantity
mode = df['Quantidade'].value_counts().index[0]

A __moda__ em Estatística é uma medida de tendência central que representa o valor mais frequente em um conjunto de dados. 

A moda é especialmente útil quando queremos saber qual é o valor mais comum ou popular em um conjunto de dados, seja em uma distribuição unimodal ou em uma distribuição bimodal. 

No entanto, a moda pode não ser representativa quanto outras medidas de tendência central, como a média e a mediana, especialmente em distribuições assimétricas ou quando há valores extremos (outliers). Por essa razão é importante analisar diferentes medidas de tendência central e usar a que melhor se adequa aos objetivos da análise estatística.

In [11]:
# Imprimindo o valor da moda (o valor que apareceu mais vezes)
print(mode)

3.0


In [17]:
# Preenchendo os valores NaN da coluna Quantidade com a moda
df.fillna({'Quantidade': mode}, inplace = True)

In [19]:
# Verificando se a quantidade de NaN
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      0
dtype: int64

`Observação:` Usar o `inplace=True` garante que a alteraçãp seja feita no dataset original e não em uma cópia do dataframe.

### Query de Dados no DataFrame do Pandas
Com o Pandas criamos dataframes que são essencialmente tabelas. Como tal, podemos fazer consultas, ou simplesmentes queries. E para isso usamos o método `query()`.

In [20]:
df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [21]:
# Checamos os valores mínimo e máximo da coluna Valor_Venda
df.Valor_Venda.describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

O intervalo de vendas __Valor_Venda__ é de 0,44 a 22638. Vamos fazer uma consulta e retornar as vendas dentro de um range de valores.

In [22]:
# Geramos um novo df aapeas com o intervalor de vendas entre 229 e 10000
df2 = df.query('229 < Valor_Venda < 10000')

In [23]:
df2.Valor_Venda.describe()

count    2357.000000
mean      766.679142
std       856.315136
min       229.544000
25%       323.100000
50%       490.320000
75%       859.200000
max      9892.740000
Name: Valor_Venda, dtype: float64

In [25]:
# Geramos um novo df apenas com os valores acima da média
df3 = df2.query('Valor_Venda > 766')

In [26]:
df3.head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
3   US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
7   CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
10  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
11  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
24  CA-2015-106320  2015-09-25   EB-13870  Consumer  United States   West   

         ID_Produto   Categoria  \
3   FUR-TA-10000577   Furniture   
7   TEC-PH-10002275  Technology   
10  FUR-TA-10001539   Furniture   
11  TEC-PH-10002033  Technology   
24  FUR-TA-10000577   Furniture   

                                     Nome_Produto  Valor_Venda  Quantidade  
3   Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
7                  Mitel 5320 IP Phone VoIP phone     907.1520         6.0  
10       Chromcraft Rectangular Conference Tables    1706.1840         9.0  
11  Konftel 250 Conference phone - Charcoal black     911.4240         4.0  
24  Bretford CR4500 Series Slim Rectangular Table    1044.6300         3.0

### Verificando a Ocorrência de Diversos Valores em Uma Coluna

Em nosso conjunto de dados do exemplo temos a coluna Quantidade que representa a quantidade de itens vendidos em cada uma das vendas. Imagine que precisamos saber em quais vendas foram vendidos 5, 7, 9 ou 11 itens. O Pandas oferece o método `isin()`para checar diversos valores em uma coluna. Esse método tem a mesma função do __IN__ presente na linguagem SQL.

In [29]:
# Retornando o número de linhas e colunas
df.shape

(9994, 11)

In [30]:
df[df['Quantidade'].isin([5, 7, 9, 11])]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14    US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
...              ...         ...        ...          ...            ...   
9974  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9977  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9981  CA-2017-163566  2017-08-03   TB-21055     Consumer  United States   
9982  US-2016-157728  2016-09-22   RC-19960     Consumer  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
5        West  FUR-FU-10001487        Furniture   
9        West  OFF-AP-10002892  Office Supplies   
10       West  FUR-TA-10001539        Furniture   
14    Central  OFF-AP-10002311  Office Supplies   
...       ...              ...              ...   
9974     West  OFF-AR-10004752  Office Supplies   
9977     West  OFF-FA-10003467  Office Supplies   
9981     East  OFF-LA-10004484  Office Supplies   
9982  Central  OFF-PA-10002195  Office Supplies   
9988    South  TEC-PH-10004006       Technology   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
5     Eldon Expressions Wood and Plastic Desk Access...      48.8600   
9                      Belkin F5C206VTEL 6 Outlet Surge     114.9000   
10             Chromcraft Rectangular Conference Tables    1706.1840   
14    Holmes Replacement Filter for HEPA Air Cleaner...      68.8100   
...                                                 ...          ...   
9974                               Blackstonian Pencils      18.6900   
9977           Alliance Big Bands Rubber Bands, 12/Pack      13.8600   
9981                                          Avery 476      16.5200   
9982  RSVP Cards & Envelopes, Blank White, 8-1/2" X ...      35.5600   
9988                    Panasonic KX - TS880B Telephone     206.1000   

      Quantidade  
3            5.0  
5            7.0  
9            5.0  
10           9.0  
14           5.0  
...          ...  
9974         7.0  
9977         7.0  
9981         5.0  
9982         7.0  
9988         5.0  

[2128 rows x 11 columns]

In [31]:
# Shape 
df[df['Quantidade'].isin([5, 7, 9, 11])].shape

(2128, 11)

In [33]:
# Mostrando as 10 primeiras linhas
df[df['Quantidade'].isin([5, 7, 9, 11])][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
21  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
22  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
27  US-2015-150630  2015-09-17   TB-21520     Consumer  United States   
35  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
36  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   

     Regiao       ID_Produto        Categoria  \
3     South  FUR-TA-10000577        Furniture   
5      West  FUR-FU-10001487        Furniture   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
14  Central  OFF-AP-10002311  Office Supplies   
21  Central  OFF-AR-10000246  Office Supplies   
22  Central  OFF-AP-10001492  Office Supplies   
27     East  FUR-BO-10004834        Furniture   
35  Central  TEC-PH-10004977       Technology   
36  Central  FUR-FU-10003664        Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
5   Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0  
21                                         Newell 318      19.4600         7.0  
22        Acco Six-Outlet Power Strip, 4' Cord Length      60.3400         7.0  
27  Riverside Palais Royal Lawyers Bookcase, Royal...    3083.4300         7.0  
35                                        GE 30524EE4    1097.5440         7.0  
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200         5.0